### ISSUE TITLE
19 samples failed in reproducing PDB timesplit set

### ISSUE THREAD

**Issue by BlacksheepJane** [CAODH/SurfDock#41]
Hi, thank you for your great work on SurfDock!

While reproducing the evaluation on the PDB timesplit (363) dataset using `eval_samples.sh`, I encountered failures in docking for 19 molecular complexes. 

The PDB IDs along with the failure reasons are as follows:

| PDB ID | Error Message                                      |
|--------|--------------------------------------------------|
| 6qlp   | index 284 is out of bounds for axis 0 with size 284  |
| 6d08   | index 272 is out of bounds for axis 0 with size 272  |
| 6qtr   | Sizes of tensors must match except in dimension 1   |
| 6cf7   | Sizes of tensors must match except in dimension 1   |
| 6jam   | index 165 is out of bounds for axis 0 with size 165  |
| 6pnn   | Sizes of tensors must match except in dimension 1   |
| 6oim   | Sizes of tensors must match except in dimension 1   |
| 6k2n   | index 257 is out of bounds for axis 0 with size 257  |
| 6qi7   | Sizes of tensors must match except in dimension 1   |
| 6olx   | Sizes of tensors must match except in dimension 1   |
| 6pya   | Sizes of tensors must match except in dimension 1   |
| 6jag   | Sizes of tensors must match except in dimension 1   |
| 6qr7   | Sizes of tensors must match except in dimension 1   |
| 6cjj   | Sizes of tensors must match except in dimension 1   |
| 6h14   | index 168 is out of bounds for axis 0 with size 168  |
| 6i78   | index 105 is out of bounds for axis 0 with size 105  |
| 6jao   | index 160 is out of bounds for axis 0 with size 160  |
| 6nd3   | Sizes of tensors must match except in dimension 1   |
| 6kjf   | Sizes of tensors must match except in dimension 1   |

From the above, there appear to be two main categories of errors:

1. `index XXX is out of bounds for axis 0 with size XXX`  
2. `Sizes of tensors must match except in dimension 1.`

After preliminary investigation, I found that both issues originate in the `get_complex` function within `SurfDock/score_in_place_dataset/score_dataset.py`.

- The first issue occurs at the line:  
  ```python
  mda_rec_model = mda.Universe(pure_pocket_path)

- The second issue occurs during the call:
  ```python
  get_rec_graph(mda_rec_model, rec_coords, c_alpha_coords, n_coords, c_coords, complex_graph, rec_radius=self.receptor_radius,
              c_alpha_max_neighbors=self.c_alpha_max_neighbors, all_atoms=self.all_atoms,
              atom_radius=self.atom_radius, atom_max_neighbors=self.atom_max_neighbors, remove_hs=self.remove_hs, lm_embeddings=lm_embeddings)

May I ask if you have encountered similar issues when testing on this dataset? If so, could you share how you addressed them?

I would really appreciate any help or guidance you can provide. Thanks again for the great work!

**Comment by BlacksheepJane** [comment]
I conducted tests on the above issue and found that it was caused by the presence of two `END` entries in the `XXX_pure.pdb` file. After removing the content after the first `END`, the first issue was resolved: three docking cases succeeded, while the remaining ones failed, leading to what I now refer to as the second issue.

The second issue arises in the `get_calpha_graph` function in `SurfDock/datasets/process_mols.py`, where the shapes of `node_feat` and `lm_embeddings` do not match in the first dimension. For example:

```python
2025-07-23 17:25:37.918 | INFO | datasets.process_mols:get_calpha_graph:515 - [node_feat.shape] torch.Size([13, 10])
2025-07-23 17:25:37.918 | INFO | datasets.process_mols:get_calpha_graph:516 - [lm_embeddings.shape] torch.Size([36, 1280])
```

This mismatch directly causes the following line to raise an error:

```python
complex_graph['receptor'].x = torch.cat([node_feat, lm_embeddings], axis=1) if lm_embeddings is not None else node_feat
```
Error:

```python
Sizes of tensors must match except in dimension 1.
```

Do you have any ideas on how to resolve this issue?

**Comment by CAODH** [comment]
> I conducted tests on the above issue and found that it was caused by the presence of two `END` entries in the `XXX_pure.pdb` file. After removing the content after the first `END`, the first issue was resolved: three docking cases succeeded, while the remaining ones failed, leading to what I now refer to as the second issue.
> 
> The second issue arises in the `get_calpha_graph` function in `SurfDock/datasets/process_mols.py`, where the shapes of `node_feat` and `lm_embeddings` do not match in the first dimension. For example:
> 
> 2025-07-23 17:25:37.918 | INFO | datasets.process_mols:get_calpha_graph:515 - [node_feat.shape] torch.Size([13, 10])
> 2025-07-23 17:25:37.918 | INFO | datasets.process_mols:get_calpha_graph:516 - [lm_embeddings.shape] torch.Size([36, 1280])
> This mismatch directly causes the following line to raise an error:
> 
> complex_graph['receptor'].x = torch.cat([node_feat, lm_embeddings], axis=1) if lm_embeddings is not None else node_feat
> Error:
> 
> Sizes of tensors must match except in dimension 1.
> Do you have any ideas on how to resolve this issue?

I did encounter this issue before, but I can't recall the exact solution. Could you provide specific error details or logs? I'll conduct immediate tests and follow up promptly

**Comment by BlacksheepJane** [comment]
> > I conducted tests on the above issue and found that it was caused by the presence of two `END` entries in the `XXX_pure.pdb` file. After removing the content after the first `END`, the first issue was resolved: three docking cases succeeded, while the remaining ones failed, leading to what I now refer to as the second issue.
> > The second issue arises in the `get_calpha_graph` function in `SurfDock/datasets/process_mols.py`, where the shapes of `node_feat` and `lm_embeddings` do not match in the first dimension. For example:
> > 2025-07-23 17:25:37.918 | INFO | datasets.process_mols:get_calpha_graph:515 - [node_feat.shape] torch.Size([13, 10])
> > 2025-07-23 17:25:37.918 | INFO | datasets.process_mols:get_calpha_graph:516 - [lm_embeddings.shape] torch.Size([36, 1280])
> > This mismatch directly causes the following line to raise an error:
> > complex_graph['receptor'].x = torch.cat([node_feat, lm_embeddings], axis=1) if lm_embeddings is not None else node_feat
> > Error:
> > Sizes of tensors must match except in dimension 1.
> > Do you have any ideas on how to resolve this issue?
> 
> I did encounter this issue before, but I can't recall the exact solution. Could you provide specific error details or logs? I'll conduct immediate tests and follow up promptly

Hi，Apologies for the delayed response, I had some matters to attend to over the past few days.

The error occurs during`get_rec_graph` in the `get_complex` function, using ligand from pdb id 6i78 as example.

Here’s a relevant excerpt from the log output:
```
2025-08-07 21:19:45.621 | INFO     | score_in_place_dataset.score_dataset:get_complex:171 - Processing 6i78
2025-08-07 21:19:45.695 | INFO     | datasets.process_mols:get_calpha_graph:515 - [node_feat.shape] torch.Size([13, 10])
2025-08-07 21:19:45.695 | INFO     | datasets.process_mols:get_calpha_graph:516 - [lm_embeddings.shape] torch.Size([36, 1280])
2025-08-07 21:19:45.695 | INFO     | datasets.process_mols:get_calpha_graph:517 - [num residues] <ResidueGroup [<Residue VAL, 116>, <Residue PRO, 117>, <Residue ASN, 143>, ..., <Residue ASN, 160>, <Residue PRO, 161>]>
2025-08-07 21:19:45.699 | INFO     | score_in_place_dataset.score_dataset:get_complex:197 - [6i78] get_rec_graph failed: Sizes of tensors must match except in dimension 1. Expected size 13 but got size 36 for tensor number 1 in the list.
2025-08-07 21:19:45.699 | INFO     | score_in_place_dataset.score_dataset:get_complex:204 - Skipping 6i78 because of the error:'NoneType' object has no attribute 'shape'
```
As you can see, I added three logging lines in `get_calpha_graph` for better insight:
```
logger.info(f"[node_feat.shape] {node_feat.shape}")
logger.info(f"[lm_embeddings.shape] {lm_embeddings.shape}")
logger.info(f"[num residues] {mdn_rec.residues}")
```
It seems like there is a mismatch between the number of residues used to compute node_feat (13) and the number of residues represented in the ESM embedding (lm_embeddings.shape[0] = 36), which leads to the downstream tensor mismatch and failure.

Please let me know if you need further details, or if there’s anything I can help with in terms of debugging or checking preprocessing assumptions.
